### ISSUE TITLE
Window May Fail to Get Keyboard Focus

### ISSUE THREAD

**Issue by hftsai256** [hyprwm/Hyprland#5764]
### Hyprland Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch  at commit fe7b748eb668136dd0558b7c8279bfcd7ab4d759  (props: bump version to 0.39.1).
Date: Tue Apr 16 16:01:03 2024
Tag: v0.39.1, commits: 4460

flags: (if any)


System Information:
System name: Linux
Node name: cyt-ht-9315
Release: 6.6.28-1-lts
Version: #1 SMP PREEMPT_DYNAMIC Wed, 17 Apr 2024 10:11:09 +0000


GPU information: 
0000:00:02.0 VGA compatible controller [0300]: Intel Corporation Alder Lake-UP4 GT2 [Iris Xe Graphics] [8086:46aa] (rev 0c) (prog-if 00 [VGA controller])


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:
```

</details>

### Bug or Regression?

Bug

### Description

Occasionally, when switching focus between multiple windows, the most recent focused window may not be gaining keyboard focus. All keyboard input will be directed to the previously focused one.

### How to reproduce

I have tested and tried to reproduce this bug with three or more windows appearing on the same screen: a terminal (Kitty, Foot), a graphical text editor (Kate), and a browser (Brave, Firefox). Set `follow_mouse = 1 or 2` and switch focus rapidly. Most of the time, the terminal window is the easiest one to trigger this bug.

### Crash reports, logs, images, videos

Launching Foot from Kitty finally shows something useful. The following warning shows up if and only if the bug is triggered.
```
warn: ime.c:65: compositor sent ime::enter() event before the corresponding keyboard_enter() event
```
The triggering rate is about 50/50. Possibly racing condition?

**Comment by akeauoupolee** [comment]
I had this problem too, and at first I thought it was due to my window rule configuration. After I commented out the window rule, the problem persisted

**Comment by hekatonkheirex** [comment]
Happens to me as well. Version 0.39

**Comment by odlove** [comment]
Happens to me as well

**Comment by The-Lost-Light** [comment]
May be this problem (dowgrade fcitx5 to 5.1.8 can temporary fix it)
[#fcitx/fcitx5/issues/1031](https://github.com/fcitx/fcitx5/issues/1031)

**Comment by stanly0726** [comment]
> May be this problem (dowgrade fcitx5 to 5.1.8 can temporary fix it) [#fcitx/fcitx5/issues/1031](https://github.com/fcitx/fcitx5/issues/1031)

i have similar problems
i found out that if i use keybind to move focus around(super+arrow keys)
if i do `SUPER+Right Arrow` and then release all keys then `SUPER+Left Arrow` then release all keys, the problem doesn't exist no matter the speed i switch, but if i hold down `SUPER` and then tapping left and right arrow key repeatedly the terminal only accept keyboard input every other focus

**Comment by clsty** [comment]
By using `git bisect`, this commit of fcitx5 is the fisrt commit that causes this issue:

https://github.com/fcitx/fcitx5/commit/b2924bd361680c493463d240a375b3f0948ae48d

PS. [here](https://github.com/fcitx/fcitx5/issues/1031#issuecomment-2080231237) fcitx5 dev had said :

> There may be some change triggered a bug in hyprland, but it's definitely not a fcitx bug.

**Comment by vaxerski** [comment]
yes, that's very possible as I've rewritten tiv3 recently, but the guy in #5763 didnt mention IMEs... anyways dupe, let's move there

**Comment by milgner** [comment]
This issue completely makes Hyprland unusable for me. I even typed a password into the chat because 1password was shown as having focus but keyboard input was routed to the browser instead.
I'll have to switch back to Xorg and i3 for now.
